src/lib/WIFI/lr1121.cpp

```diff
@@ -0,0 +1,233 @@
+    SPIEx.transferBytes(lr1121UpdateState->packet.header, nullptr, 6 + write_size);
+    digitalWrite(lr1121UpdateState->updatingRadio == SX12XX_Radio_1 ? GPIO_PIN_NSS : GPIO_PIN_NSS_2, HIGH);
+
+    while (digitalRead(lr1121UpdateState->updatingRadio == SX12XX_Radio_1 ? GPIO_PIN_BUSY : GPIO_PIN_BUSY_2) == HIGH)
+    {
+        delay(1);
+    }
+    lr1121UpdateState->totalSize += write_size;
+    lr1121UpdateState->left_over = 0;
+    DBGLN("flashed");
+}
+
+static void readRegister(SX12XX_Radio_Number_t radio, uint16_t reg, uint8_t *buffer, uint32_t buffer_len)
+{
+    uint8_t read[5];
+    read[0] = reg >> 8;
+    read[1] = (uint8_t)reg;
+    SPIEx.write(radio, read, 2);
+    hal.WaitOnBusy(radio);
+    memset(buffer, 0, buffer_len);
+    SPIEx.read(radio, buffer, buffer_len);
+    hal.WaitOnBusy(radio);
+}
+
+static void WebUploadLR1121ResponseHandler(AsyncWebServerRequest *request) {
+    // Complete upload and set error flag
+    bool uploadError = false;
+    uint8_t param = 0;
+    writeLR1121Bytes(nullptr, 0);
+    lr1121UpdateState->totalSize += lr1121UpdateState->left_over;
+
+    SPIEx.setHwCs(true);
+    if (GPIO_PIN_NSS_2 != UNDEF_PIN)
+    {
+        spiAttachSS(SPIEx.bus(), 1, GPIO_PIN_NSS_2);
+    }
+
+    if (lr1121UpdateState->totalSize == lr1121UpdateState->expectedFilesize)
+    {
+        DBGLN("reboot 1121");
+        uint8_t reboot_cmd[] = {
+            (uint8_t)(LR11XX_BL_REBOOT_OC >> 8),
+            (uint8_t)LR11XX_BL_REBOOT_OC,
+            0
+        };
+        SPIEx.write(lr1121UpdateState->updatingRadio, reboot_cmd, 3);
+        while(!hal.WaitOnBusy(lr1121UpdateState->updatingRadio));
+
+        DBGLN("check not in BL mode");
+        uint8_t packet[5];
+        readRegister(lr1121UpdateState->updatingRadio, LR11XX_SYSTEM_GET_VERSION_OC, packet, 5);
+        uploadError = (packet[2] != 3);
+        DBGLN("hardware %x", packet[1]);
+        DBGLN("type %x", packet[2]);
+        DBGLN("firmware %x", ( ( uint16_t )packet[3] << 8 ) + ( uint16_t )packet[4]);
+    }
+
+    String msg;
+    if (!uploadError && lr1121UpdateState->totalSize == lr1121UpdateState->expectedFilesize) {
+        msg = String(R"({"status": "ok", "msg": "Update complete. Refresh page to see new version information."})");
+        DBGLN("Update complete");
+    } else {
+        StreamString p = StreamString();
+        if (lr1121UpdateState->totalSize != lr1121UpdateState->expectedFilesize) {
+            p.println("Not enough data uploaded!");
+        } else {
+            p.println("Update failed, refresh and try again.");
+        }
+        DBGLN("Failed to upload firmware: %s", p.c_str());
+        msg = String(R"({"status": "error", "msg": ")") + p + R"("})";
+    }
+    AsyncWebServerResponse *response = request->beginResponse(200, "application/json", msg);
+    response->addHeader("Connection", "close");
+    request->send(response);
+    request->client()->close();
```

Comment:
Missing `delete lr1121UpdateState; lr1121UpdateState = nullptr`?

Also curious is it necessary to force close the connection? Usually we do this when we're going to reboot but in this case it seems unnecessary.

Reply:
Damn it! I thought I had covered all my bases.